File: spring-boot-request-validation/src/main/java/com/github/erosb/kappa/examples/PetController.java
package com.github.erosb.kappa.examples;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.LocalDate;

class UserIdentifier {
    String id;
}

class CreatePetRequest{
    String name;
    UserIdentifier owner;
    LocalDate birthDate;
}

@RestController
@RequestMapping("/api/pets")
public class PetController {

    @PostMapping
    void createPet(@RequestBody CreatePetRequest requestBody) {
        System.out.println("requestBody = " + requestBody);
    }
}

File: spring-boot-validation-example/src/main/kotlin/com/github/erosb/openapidemo/PetController.kt
package com.github.erosb.openapidemo

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate


data class UserIdentifier(
    var id: Int
)

data class CreatePetRequest(
    var name: String,
    var owner: UserIdentifier,
    var birthDate: LocalDate
)

@RestController
@RequestMapping("/api/pet")
class PetController {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun createPet(@RequestBody request: CreatePetRequest): Any {
        return request
    }
}
